CONTRIB.md

```diff
@@ -15,24 +15,59 @@
 
 # Apache NiFi - MiNiFi - C++ Contribution Guide
 
-
-We welcome all contributions to Apache MiNiFi. To make development easier, we've included
-the linter for the Google Style guide. Google provides an Eclipse formatter for their style
-guide. It is located [here](https://github.com/google/styleguide/blob/gh-pages/eclipse-cpp-google-style.xml).
-New contributions are expected to follow the Google style guide when it is reasonable.
-Additionally, all new files must include a copy of the Apache License Header.
-
+We welcome all contributions to Apache MiNiFi. All new files must include a copy of the Apache License Header.
+To make development easier, we've included the linter for the Google Style guide. Google provides an Eclipse formatter
+for their style guide. It is located
+[here](https://github.com/google/styleguide/blob/gh-pages/eclipse-cpp-google-style.xml).
+New contributions are expected to follow the Google Style Guide, except for the following points:
+- Use .cpp extension for implementation files
+- Use lowerCamelCase for functions, including accessors/mutators
+- Use UPPER_SNAKE_CASE for constants
+- Filenames are typically class names or a description of the contents in UpperCamelCase
+- If a class is imitating something from STL, boost or similar, then STL-style lower_snake_case is used for naming the
+  class. UpperSnakeCase is used for most classes, in line with the Google Style Guide.
+- Prefer `#pragma once` over include guards
+- Forward declarations are OK
+- Using-directives (`using namespace foo`) are discouraged, except for user-defined literal namespaces
+- Some patterns in the codebase rely on objects with static storage duration without being trivially destructible and
+  initialized with a constant expression. It's OK to use these.
+- Operator overloading and user-defined literal suffixes are OK
```

Comment:
Suggested addition:
```suggestion
- Operator overloading and user-defined literal suffixes are OK, but use them sparingly and only where they make the code more readable
```

Reply:
I don't want to discourage their usage. Using operator-overloading even for syntactic sugar is fine by my standards, it makes great interfaces like [Clara](https://github.com/catchorg/Clara) possible. And UDLs can also be helpful, see Literals.h in our codebase or the [units](https://github.com/nholthaus/units) library. Of course they can be misused, but I'd rather catch those during reviews than to potentially prevent their benefits from being utilized because of the discouraging wording.

Reply:
Operator overloading can be nice, especially if done well in a library.  But it can easily be done poorly, and it does make navigating the code harder.  So I would like to lightly discourage it, as above, but I'd be interested in what others think.

Reply:
I'm not a fan of operator overloading, but it is because of my own bad experiences. But I think my own bad experiences shouldn't discourage anyone who would create an excellent interface through operator overloading. I wouldn't add the `but use them sparingly` part, and I would rather say `but only use them if they do not make the code less readable`. In that case if we catch a similar problem on review we can point to this part of the CONTRIB.md.

Reply:
What do you think about not adding anything here and adding a sentence in the text part, emphasizing simplicity and readability? Like "Contributions should aim for simple and readable code".

Reply:
I'm okay with that.

Reply:
I came up with a slightly different version in 3fd87be. I'm hoping for to reach consensus, let me know if you like it or not.

Reply:
I would still prefer some indication in this section that operator overloading can easily result in those later-mentioned footguns, but I'm OK with the current version if everyone else is.

Reply:
How about removing the Operator overloading part, falling back on the [style guide](https://google.github.io/styleguide/cppguide.html#Operator_Overloading), which says "Overload operators judiciously"? It also disallows overloading &&, ||, , (comma), and unary &, which seems fine to me, and there is always a possible fallback to this in extra special cases: "It's ok to diverge from any of the rules with a good enough reason.".
```suggestion
- User-defined literal suffixes are OK
```